syntax = "proto3";

package geograf;

message Lat {
  double value = 1;
}

message Lon {
  double value = 1;
}

message GpsPoint {
  Lat latitude = 1;
  Lon longitude = 2;
}

message MatchRequest {
  repeated GpsPoint recorded_route = 1;
  repeated GpsPoint predefined_route = 2;
}

message MatchResponse {
  oneof result {
    MatchSuccess success = 1;
    MatchError error = 2;
  }
}

message CumulativeDistance {
  double meters = 1;
}

message MatchSuccess {
  repeated GpsPoint matched_points = 1;
  repeated CumulativeDistance distances = 2;
}

message MatchError {
  ErrorCode code = 1;
  string details = 2;
}

enum ErrorCode {
  NO_MATCH = 0;
}

message Lap {
  repeated GpsPoint points = 1;
}

message LapDistanceRequest {
  Lap lap = 1;
}

message LapDistanceResponse {
  repeated CumulativeDistance distances = 1;
}

message TimeSplit {
  GpsPoint from = 1;
  GpsPoint to = 2;
}

message TimesplitOnLapRequest {
  Lap lap = 1;

  // produced by LapDistanceResponse
  repeated CumulativeDistance distances = 2;

  TimeSplit timesplit = 3;
}

message TimesplitOnLap {
  GpsPoint intersection_point = 1;
  // Cumulative distance along the lap at the intersection point
  CumulativeDistance distance_on_lap = 2;
}

message TimesplitOnLapResponse {
  repeated TimesplitOnLap timesplits_on_lap = 1;
}

service Geograf {
  rpc MatchRoutes (MatchRequest) returns (MatchResponse);

  rpc CalculateLapDistance (LapDistanceRequest) returns (LapDistanceResponse);

  // Find intersections between a timesplit (line) and a lap polyline.
  // Returns list of intersections with the GPS point and its cumulative distance on the lap.
  rpc FindTimesplitOnLap (TimesplitOnLapRequest) returns (TimesplitOnLapResponse);
}